.azure-pipelines/client.yml

```diff
@@ -18,45 +18,113 @@ jobs:
 
   - script: |
      pip install wheel
+     pip install setuptools
+     pip install pathlib
+    displayName: 'Prep Environment'
 
-    displayName: 'Prep Environment, Install Packages'
-
-  - powershell: |
-     $inputFilterSet="$(build_targeting_string)"
-     
-     if ($inputFilterSet -eq "")
-     {
-         $inputFilterSet = "azure-*"
-     }
-     
-     foreach($inputFilter in $inputFilterSet.split(","))
-     {
-       $foldersToIterate = gci -Filter "$inputFilter"
-     
-       foreach ($folder in $foldersToIterate)
-       {
-         python3 ./build_package.py --dest "$(Build.ArtifactStagingDirectory)" "$folder"
-       }
-     }
-     
+  - task: PythonScript@0
     displayName: 'Generate Packages'
+    inputs:
+      scriptPath: 'scripts/devops_tasks/build_packages.py'
+      arguments: '-d "$(Build.ArtifactStagingDirectory)" "$(build_targeting_string)"'
 
   - task: PublishBuildArtifacts@1
     displayName: 'Publish Artifacts'
 
-- template: generic.test.yml
-  parameters:
-    name: Linux
+- job: 'Generic_Tests'
+  dependsOn:
+    - 'Build_Target_Packages'
+  strategy:
+    matrix:
+      Linux_Python27:
+        os.name: 'Linux'
+        os.vmImage: 'ubuntu-16.04'
+        python.version: '2.7'
+      Linux_Python34:
+        os.name: 'Linux'
+        os.vmImage: 'ubuntu-16.04'
+        python.version: '3.4'
+      Linux_Python35:
+        os.name: 'Linux'
+        os.vmImage: 'ubuntu-16.04'
+        python.version: '3.5'
+      Linux_Python36:
+        os.name: 'Linux'
+        os.vmImage: 'ubuntu-16.04'
+        python.version: '3.6'
+      Linux_Python37:
+        os.name: 'Linux'
+        os.vmImage: 'ubuntu-16.04'
+        python.version: '3.7'
+      Windows_Python35:
+        os.name: 'Windows'
+        os.vmImage: 'vs2017-win2016'
+        python.version: '3.5'
+      MacOS_Python27:
+        os.name: 'MacOS'
+        os.vmImage: 'macOS-10.13'
+        python.version: '2.7'
+  pool:
+    vmImage: '$(os.vmImage)'
+
+  steps:
+  - task: UsePythonVersion@0
+    displayName: 'Use Python $(python.version)'
+    inputs:
+      versionSpec: '$(python.version)'
+
+  - script: |
+     pip install wheel
+     pip install setuptools
+     pip install pathlib
+     pip install twine
+    displayName: 'Prep Environment'
+
+  - task: PythonScript@0
+    displayName: 'Setup and Run Tests'
+    inputs:
+      scriptPath: 'scripts/devops_tasks/setup_execute_tests.py'
+      arguments: '"$(build_targeting_string)"'
+
+- job: Special_Python_Distro_Tests
+  dependsOn:
+  - 'Build_Target_Packages'
+  pool:
     vmImage: 'ubuntu-16.04'
+  steps:
+  - script: |
+     pip install wheel
+     pip install setuptools
+     pip install pathlib
+    displayName: 'Prep Environment'
 
-- template: generic.test.yml
-  parameters:
-    name: macOS
-    vmImage: 'macOS-10.13'
+  - script: |
+     cd ~/
+     wget https://bitbucket.org/pypy/pypy/downloads/pypy3-v6.0.0-linux64.tar.bz2
```

Comment:
Is bitbucket the best place to download pypy?

Reply:
It is where their official site points to.